File: python/src/orderflow_metrics/downsidecov.py
"""Downside covariance and correlation matrices - the crash-correlation half of
the covariance matrix, for a whole book of assets.

``semicovariance`` splits the realized covariance of *two* series by the sign of
their returns. This lifts the joint-downside piece to a full N x N matrix: for
every pair it keeps only the days both assets fell::

    D_ij = sum min(x_i,t, 0) * min(x_j,t, 0)

- the concordant-negative (both-down) component of the Bollerslev, Li, Patton &
Quaedvlieg (2020) decomposition. The diagonal ``D_ii = sum min(x_i,0)**2`` is
asset i's downside semivariance (see ``semivar``), so normalizing gives a
downside *correlation* matrix. Its off-diagonal average is a single, monitorable
read on how tightly a book is bound together on the way down - the "correlations
go to one in a crisis" effect turned into a number.

All series are paired by index and truncated to their common length, so align
them to the same sampling grid first. An empty input returns an empty matrix.
"""
from __future__ import annotations

import math
from typing import List, Sequence


def _common_length(series: Sequence[Sequence[float]]) -> int:
    if not series:
        return 0
    return min(len(s) for s in series)


def downside_covariance_matrix(
    series: Sequence[Sequence[float]],
) -> List[List[float]]:
    """N x N downside covariance matrix: entry ``[i][j]`` is the joint-downside
    covariance ``sum min(x_i,0)*min(x_j,0)`` over the two series' common length -
    the both-down component of realized semicovariance. Symmetric and positive
    semidefinite; the diagonal is each asset's downside semivariance. Returns
    ``[]`` for empty input.
    """
    n = len(series)
    t = _common_length(series)
    neg = [[(s[k] if s[k] < 0 else 0.0) for k in range(t)] for s in series]
    m = [[0.0] * n for _ in range(n)]
    for i in range(n):
        for j in range(i, n):
            total = 0.0
            for k in range(t):
                total += neg[i][k] * neg[j][k]
            m[i][j] = total
            m[j][i] = total
    return m


def downside_correlation_matrix(
    series: Sequence[Sequence[float]],
) -> List[List[float]]:
    """N x N downside correlation matrix: the downside covariance matrix
    normalized by the square root of its diagonal, ``D_ij / sqrt(D_ii * D_jj)``.
    The diagonal is 1 (or ``nan`` for an asset with no downside variance). Any
    pair involving a zero-downside asset is ``nan``. Returns ``[]`` for empty
    input.
    """
    d = downside_covariance_matrix(series)
    n = len(d)
    r = [[math.nan] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            denom = d[i][i] * d[j][j]
            r[i][j] = d[i][j] / math.sqrt(denom) if denom > 0 else math.nan
    return r


def average_downside_correlation(series: Sequence[Sequence[float]]) -> float:
    """Average off-diagonal downside correlation: the mean of the upper-triangle
    entries of the downside correlation matrix - one number for how correlated a
    book is on the way down. ``nan`` pairs (zero-downside assets) are skipped;
    returns ``nan`` when no valid pair exists (fewer than two assets, or none
    share downside variance).
    """
    r = downside_correlation_matrix(series)
    n = len(r)
    total = 0.0
    count = 0
    for i in range(n):
        for j in range(i + 1, n):
            if not math.isnan(r[i][j]):
                total += r[i][j]
                count += 1
    return math.nan if count == 0 else total / count

File: src/downsidecov.ts
/**
 * Downside covariance and correlation matrices — the crash-correlation half of
 * the covariance matrix, for a whole book of assets.
 *
 * `semicovariance.ts` splits the realized covariance of *two* series by the sign
 * of their returns. This lifts the joint-downside piece to a full N×N matrix:
 * for every pair it keeps only the days both assets fell,
 *
 *   D_ij = Σ min(xᵢ,ₜ, 0) · min(xⱼ,ₜ, 0)
 *
 * — the concordant-negative (both-down) component of the Bollerslev, Li, Patton
 * & Quaedvlieg (2020) decomposition. The diagonal `D_ii = Σ min(xᵢ,0)²` is asset
 * i's downside semivariance (see `semivar.ts`), so normalizing gives a downside
 * *correlation* matrix. Its off-diagonal average is a single, monitorable read
 * on how tightly a book is bound together on the way down — the &ldquo;correlations
 * go to one in a crisis&rdquo; effect turned into a number.
 *
 * All series are paired by index and truncated to their common length, so align
 * them to the same sampling grid first. An empty input returns an empty matrix.
 */

/** Common length across every series (0 if none). */
function commonLength(series: readonly (readonly number[])[]): number {
  if (series.length === 0) return 0;
  let t = Infinity;
  for (const s of series) t = Math.min(t, s.length);
  return t === Infinity ? 0 : t;
}

/**
 * N×N downside covariance matrix: entry `[i][j]` is the joint-downside
 * covariance `Σ min(xᵢ,0)·min(xⱼ,0)` over the two series' common length — the
 * both-down component of realized semicovariance. Symmetric and positive
 * semidefinite; the diagonal is each asset's downside semivariance. Returns `[]`
 * for empty input.
 */
export function downsideCovarianceMatrix(
  series: readonly (readonly number[])[],
): number[][] {
  const n = series.length;
  const t = commonLength(series);
  const m: number[][] = Array.from({ length: n }, () => new Array<number>(n).fill(0));
  // negative parts, precomputed
  const neg: number[][] = series.map((s) => {
    const out = new Array<number>(t);
    for (let k = 0; k < t; k++) out[k] = s[k] < 0 ? s[k] : 0;
    return out;
  });
  for (let i = 0; i < n; i++) {
    for (let j = i; j < n; j++) {
      let sum = 0;
      for (let k = 0; k < t; k++) sum += neg[i][k] * neg[j][k];
      m[i][j] = sum;
      m[j][i] = sum;
    }
  }
  return m;
}

/**
 * N×N downside correlation matrix: the downside covariance matrix normalized by
 * the square root of its diagonal, `D_ij / √(D_ii · D_jj)`. The diagonal is 1
 * (or `NaN` for an asset with no downside variance). Any pair involving a
 * zero-downside asset is `NaN`. Returns `[]` for empty input.
 */
export function downsideCorrelationMatrix(
  series: readonly (readonly number[])[],
): number[][] {
  const d = downsideCovarianceMatrix(series);
  const n = d.length;
  const r: number[][] = Array.from({ length: n }, () => new Array<number>(n).fill(NaN));
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      const denom = d[i][i] * d[j][j];
      r[i][j] = denom > 0 ? d[i][j] / Math.sqrt(denom) : NaN;
    }
  }
  return r;
}

/**
 * Average off-diagonal downside correlation: the mean of the upper-triangle
 * entries of the downside correlation matrix — one number for how correlated a
 * book is on the way down. `NaN` pairs (zero-downside assets) are skipped;
 * returns `NaN` when no valid pair exists (fewer than two assets, or none share
 * downside variance).
 */
export function averageDownsideCorrelation(
  series: readonly (readonly number[])[],
): number {
  const r = downsideCorrelationMatrix(series);
  const n = r.length;
  let sum = 0;
  let count = 0;
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      if (!Number.isNaN(r[i][j])) {
        sum += r[i][j];
        count++;
      }
    }
  }
  return count === 0 ? NaN : sum / count;
}
